File: combo2-advanced-agentic-engineering/tiny-agent-csharp/src/TinyAgent.Starter/Agent.cs
using System.Text;
using System.Text.Json.Nodes;
using TinyAgent.Shared;

namespace TinyAgent.Starter;

/// <summary>
/// The agent loop. YOU WRITE THIS (step 1).
/// </summary>
/// <remarks>
/// <para>Thesis (Thorsten Ball, ampcode.com): <i>It's an LLM, a loop, and enough tokens.</i></para>
/// <para>The shape of the loop you're going to write:</para>
/// <code>
/// contents = [userPrompt]
/// for turn in 1..maxTurns:
///     response = await client.GenerateContentAsync(model, request)
///     contents.Add(candidate.Content)          // don't forget this line
///     calls = parts where FunctionCall is not null
///     if no calls: return the joined text      // done
///     foreach call: Dispatch it, append a functionResponse part
///     contents.Add(new Content("user", responseParts))
/// </code>
/// <para>
/// Yes, it's <c>async</c>. Every HTTP call in .NET is a <see cref="Task"/>, so
/// <c>await</c> is unavoidable. It is plumbing, not the lesson.
/// </para>
/// </remarks>
public static class Agent
{
    public const string SystemInstruction = """
        You are a careful coding assistant working inside a small
        code repository. You have three tools: read_file, list_files, and edit_file.

        Workflow:
        1. Explore first. Use list_files and read_file to build an understanding before editing.
        2. Edit sparingly. One edit per logical change. Use enough surrounding context in
           old_str so it matches exactly once.
        3. Report what you did in plain prose when you are finished. Do not call any tool on
           the final turn — that's how you signal you're done.
        4. If a tool returns a string starting with "ERROR:", read the error carefully and
           adjust your approach. Don't retry the same call blindly.
        """;

    // -------------------------------------------------------------------------
    // STEP 1 — implement RunAsync
    // -------------------------------------------------------------------------
    /// <summary>
    /// Run the agent loop until the model returns a final answer.
    /// </summary>
    /// <remarks>
    /// Hints — everything you need is in <c>TinyAgent.Shared</c>:
    /// <list type="bullet">
    /// <item>Build a turn: <c>new Content("user", new[] { new Part(Text: userPrompt) })</c></item>
    /// <item>Build the request: <c>new GeminiRequest(contents, new[] { ToolSchemas.All() }, systemInstruction)</c></item>
    /// <item>Call the model: <c>await client.GenerateContentAsync(model, request, cancellationToken)</c></item>
    /// <item>The model's turn: <c>response.Candidates?.FirstOrDefault()?.Content</c></item>
    /// <item>Function calls live in <c>content.Parts</c> where <c>part.FunctionCall is not null</c></item>
    /// <item>Send a result back: <c>new Part(FunctionResponse: new FunctionResponse(name, new JsonObject { ["result"] = result }))</c></item>
    /// <item>Tool results go back with role <b>"user"</b>, not "tool" and not "function"</item>
    /// <item>Termination: a turn with no function-call parts</item>
    /// <item><see cref="Dispatch"/> below is written for you — call it, don't rewrite it</item>
    /// </list>
    /// Start with the simplest version that handles the exploration prompts
    /// (TODO.md items 1 and 2), then try the bug-fix prompt (item 3).
    /// </remarks>
    public static Task<string> RunAsync(
        string userPrompt,
        ITools tools,
        GeminiClient client,
        string? model = null,
        int maxTurns = 20,
        Action<AgentEvent>? onEvent = null,
        CancellationToken cancellationToken = default)
    {
        model ??= DotEnv.Model();

        // TODO: Step 1 — write the loop.
        throw new NotImplementedException("Implement RunAsync for step 1.");
    }

    // ---- given below this line — no changes needed -------------------------

    /// <summary>
    /// Route one function call to a tool. Every failure path returns a string.
    /// </summary>
    public static string Dispatch(ITools tools, FunctionCall call)
    {
        try
        {
            return call.Name switch
            {
                ToolSchemas.ReadFileName => tools.ReadFile(RequiredArg(call, "path")),
                ToolSchemas.ListFilesName => tools.ListFiles(OptionalArg(call, "path") ?? ".").ToModelString(),
                ToolSchemas.EditFileName => tools.EditFile(
                    RequiredArg(call, "path"),
                    RequiredArg(call, "old_str"),
                    RequiredArg(call, "new_str")),
                _ => $"ERROR: unknown tool '{call.Name}'",
            };
        }
        catch (ArgumentException ex)
        {
            return $"ERROR: bad arguments to {call.Name}: {ex.Message}";
        }
        catch (Exception ex)
        {
            return $"ERROR: {ex.GetType().Name}: {ex.Message}";
        }
    }

    public static string RequiredArg(FunctionCall call, string name) =>
        OptionalArg(call, name) ?? throw new ArgumentException($"missing required argument '{name}'");

    public static string? OptionalArg(FunctionCall call, string name)
    {
        if (call.Args is null) return null;
        return call.Args.TryGetPropertyValue(name, out var node) && node is not null
            ? node.GetValue<string>()
            : null;
    }

    public static string PreviewArgs(JsonObject? args)
    {
        if (args is null || args.Count == 0) return string.Empty;

        var sb = new StringBuilder();
        foreach (var (key, value) in args)
        {
            if (sb.Length > 0) sb.Append(", ");
            sb.Append(key).Append('=').Append(value?.ToJsonString() ?? "null");
        }

        var preview = sb.ToString();
        return preview.Length > 120 ? preview[..117] + "..." : preview;
    }
}

File: combo2-advanced-agentic-engineering/java/tiny-agent-java/starter/src/main/java/ai/octoco/tinyagent/starter/Agent.java
package ai.octoco.tinyagent.starter;

import ai.octoco.tinyagent.shared.AgentEvent;
import ai.octoco.tinyagent.shared.DotEnv;
import ai.octoco.tinyagent.shared.FunctionCall;
import ai.octoco.tinyagent.shared.GeminiClient;
import ai.octoco.tinyagent.shared.ToolSchemas;
import ai.octoco.tinyagent.shared.Tools;
import com.fasterxml.jackson.databind.JsonNode;
import java.util.concurrent.CompletableFuture;
import java.util.function.Consumer;

/**
 * The agent loop. YOU WRITE THIS (step 1).
 *
 * <p>Thesis (Thorsten Ball, ampcode.com): <i>It's an LLM, a loop, and enough
 * tokens.</i> The shape of the loop you're going to write:
 *
 * <pre>{@code
 * contents = [userPrompt]
 * for turn in 1..maxTurns:
 *     response = client.generateContentAsync(model, request).join()
 *     contents.add(candidate.content())        // don't forget this line
 *     calls = parts where functionCall() is not null
 *     if no calls: return the joined text      // done
 *     foreach call: dispatch it, append a functionResponse part
 *     contents.add(new Content("user", responseParts))
 * }</pre>
 *
 * <p>{@link GeminiClient#generateContentAsync} returns a {@link CompletableFuture}
 * because {@code java.net.http.HttpClient.sendAsync} does. Calling {@code .join()}
 * on it is the Java equivalent of {@code await} — plumbing, not the lesson.
 */
public final class Agent {

    public static final String SYSTEM_INSTRUCTION = """
            You are a careful coding assistant working inside a small
            code repository. You have three tools: read_file, list_files, and edit_file.

            Workflow:
            1. Explore first. Use list_files and read_file to build an understanding before editing.
            2. Edit sparingly. One edit per logical change. Use enough surrounding context in
               old_str so it matches exactly once.
            3. Report what you did in plain prose when you are finished. Do not call any tool on
               the final turn — that's how you signal you're done.
            4. If a tool returns a string starting with "ERROR:", read the error carefully and
               adjust your approach. Don't retry the same call blindly.
            """;

    private Agent() {}

    // -------------------------------------------------------------------------
    // STEP 1 — implement runAsync
    // -------------------------------------------------------------------------
    /**
     * Run the agent loop until the model returns a final answer.
     *
     * <p>Hints — everything you need is in {@code ai.octoco.tinyagent.shared}:
     * <ul>
     *   <li>Build a turn: {@code new Content("user", List.of(Part.text(userPrompt)))}</li>
     *   <li>Build the request: {@code new GeminiRequest(contents, List.of(ToolSchemas.all()), systemInstruction)}</li>
     *   <li>Call the model: {@code client.generateContentAsync(model, request).join()}</li>
     *   <li>The model's turn: {@code response.firstCandidate().content()}</li>
     *   <li>Function calls live in {@code content.parts()} where {@code part.functionCall() != null}</li>
     *   <li>Send a result back: {@code Part.functionResponse(new FunctionResponse(name, resultNode))}</li>
     *   <li>Tool results go back with role <b>"user"</b>, not "tool" and not "function"</li>
     *   <li>Termination: a turn with no function-call parts</li>
     *   <li>{@link #dispatch} below is written for you — call it, don't rewrite it</li>
     * </ul>
     * Start with the simplest version that handles the exploration prompts
     * (TODO.md items 1 and 2), then try the bug-fix prompt (item 3).
     */
    public static CompletableFuture<String> runAsync(
            String userPrompt,
            Tools tools,
            GeminiClient client,
            String model,
            int maxTurns,
            Consumer<AgentEvent> onEvent) {
        String resolvedModel = model != null ? model : DotEnv.model();

        // TODO: Step 1 — write the loop.
        return CompletableFuture.failedFuture(new UnsupportedOperationException("Implement runAsync for step 1."));
    }

    // ---- given below this line — no changes needed -------------------------

    /**
     * Route one function call to a tool. Every failure path returns a string.
     */
    public static String dispatch(Tools tools, FunctionCall call) {
        try {
            return switch (call.name()) {
                case ToolSchemas.READ_FILE -> tools.readFile(requiredArg(call, "path"));
                case ToolSchemas.LIST_FILES -> {
                    String path = optionalArg(call, "path");
                    yield tools.listFiles(path != null ? path : ".").toModelString();
                }
                case ToolSchemas.EDIT_FILE -> tools.editFile(
                        requiredArg(call, "path"),
                        requiredArg(call, "old_str"),
                        requiredArg(call, "new_str"));
                default -> "ERROR: unknown tool '" + call.name() + "'";
            };
        } catch (IllegalArgumentException ex) {
            return "ERROR: bad arguments to " + call.name() + ": " + ex.getMessage();
        } catch (Exception ex) {
            return "ERROR: " + ex.getClass().getSimpleName() + ": " + ex.getMessage();
        }
    }

    public static String requiredArg(FunctionCall call, String name) {
        String value = optionalArg(call, name);
        if (value == null) throw new IllegalArgumentException("missing required argument '" + name + "'");
        return value;
    }

    public static String optionalArg(FunctionCall call, String name) {
        JsonNode args = call.args();
        if (args == null) return null;
        JsonNode node = args.get(name);
        return (node == null || node.isNull()) ? null : node.asText();
    }

    public static String previewArgs(JsonNode args) {
        if (args == null || args.isEmpty()) return "";

        StringBuilder sb = new StringBuilder();
        var fields = args.fields();
        while (fields.hasNext()) {
            var entry = fields.next();
            if (sb.length() > 0) sb.append(", ");
            sb.append(entry.getKey()).append('=').append(entry.getValue().toString());
        }

        String preview = sb.toString();
        return preview.length() > 120 ? preview.substring(0, 117) + "..." : preview;
    }
}
